src/irmin-pack/mem/content_addressable.ml

```diff
@@ -56,14 +56,22 @@ end
 module Maker (K : Irmin.Hash.S) = struct
   type key = K.t
 
-  module Make (Val : Irmin_pack.Pack_value.S with type hash := K.t) = struct
+  module Make
+      (Val : Irmin_pack.Pack_value.S
+               with type hash := K.t
+               with type key = Irmin.Key.Of_hash(K).t) =
+  struct
+    (* TODO(craigfe): We could use the keys to skip traversal of the map on
+       lookup, but this must not introduce false-positives under [clear]. *)
```

Comment:
do you have an idea how to do this?

Reply:
One strategy (the one taken by @samoht in his original PR) would be to just store the values directly in the keys where possible:

```ocaml
type key = { hash : hash; mutable cached_value : value option }
```

(This is effectively just caching the lookup of the value in the store map.) Then `find` can just check the cache inside the key before attempting anything else:

```ocaml
let find t key =
  match key.cached_value with
  | Some v -> v
  | None -> Map.find t.hashes key.hash
```

Unfortunately, there are two problems with it:

- the store map can be `clear`ed, making the cached value invalid;
- there can be multiple stores in memory at once, so any given call to `find t key` might use a `key` from a store other than `t`.

I think we could get around these problems by storing more information in the key:

```ocaml
type key =
  { hash : hash
  ; mutable cached_value : value option
  ; clear_since_cache : bool ref  (* Signal from the store that a [clear] has happened *)
  ; store_id : unit ref  (* Unique identifier for stores *)
  }

let find t key =
  if t.store_id == key.store_id && not !key.clear_since_last_cache then
    match key.cached_value with
    | Some v -> v
    | None -> Map.find t.hashes key.hash
  else
    Map.find t.hashes key.hash

let clear t =
  t.clear_flag := true;
  t.clear_flag <- ref false;
  t.hashes <- Map.empty
```

But given these complications, I'm not sure it's ever worth actually doing this. We should probably have some benchmarks for it before we attempt it, at any rate.